src/actions/clipboard.ts

```diff
@@ -0,0 +1,373 @@
+ */
+export class Clipboard {
+  /** Data copied by the copy or cut keyboard shortcuts. */
+  private copyData: ICopyData | null = null;
+
+  /** The workspace a copy or cut keyboard shortcut happened in. */
+  private copyWorkspace: WorkspaceSvg | null = null;
+
+  /**
+   * Function provided by the navigation controller to say whether editing
+   * is allowed.
+   */
+  private canCurrentlyEditFn: (ws: WorkspaceSvg) => boolean;
+
+  /**
+   * Weight for the first of these three items in the context menu.
+   * Changing base weight will change where this group goes in the context
+   * menu; changing individual weights relative to base weight can change
+   * the order within the clipboard group.
+   */
+  private baseWeight = 11;
+
+  constructor(
+    private navigation: Navigation,
+    canEditFn: (ws: WorkspaceSvg) => boolean,
+  ) {
+    this.canCurrentlyEditFn = canEditFn;
+  }
+
+  /**
+   * Install these actions as both keyboard shortcuts and context menu items.
+   */
+  install() {
+    this.registerCopyShortcut();
+    this.registerCopyContextMenuAction();
+
+    this.registerPasteShortcut();
+    this.registerPasteContextMenuAction();
+
+    this.registerCutShortcut();
+    this.registerCutContextMenuAction();
+  }
+
+  /**
+   * Uninstall this action as both a keyboard shortcut and a context menu item.
+   * Reinstall the original context menu action if possible.
+   */
+  uninstall() {
+    ContextMenuRegistry.registry.unregister('blockCopyFromContextMenu');
+    ContextMenuRegistry.registry.unregister('blockPasteFromContextMenu');
+    ContextMenuRegistry.registry.unregister('blockCutFromContextMenu');
+
+    ShortcutRegistry.registry.unregister(Constants.SHORTCUT_NAMES.COPY);
+    ShortcutRegistry.registry.unregister(Constants.SHORTCUT_NAMES.CUT);
+    ShortcutRegistry.registry.unregister(Constants.SHORTCUT_NAMES.PASTE);
+  }
+
+  /**
+   * Create and register the keyboard shortcut for the copy action.
+   */
+  private registerCopyShortcut() {
+    const copyShortcut: ShortcutRegistry.KeyboardShortcut = {
+      name: Constants.SHORTCUT_NAMES.COPY,
+      preconditionFn: this.copyPreconditionFn.bind(this),
+      callback: this.copyCallbackFn.bind(this),
+      keyCodes: [
+        createSerializedKey(KeyCodes.C, [KeyCodes.CTRL]),
+        createSerializedKey(KeyCodes.C, [KeyCodes.ALT]),
+        createSerializedKey(KeyCodes.C, [KeyCodes.META]),
+      ],
+      allowCollision: true,
+    };
+    ShortcutRegistry.registry.register(copyShortcut);
+  }
+
+  /**
+   * Register the copy block action as a context menu item on blocks.
+   * This function mixes together the keyboard and context menu preconditions
+   * but only calls the keyboard callback.
+   */
+  private registerCopyContextMenuAction() {
+    const copyAction: ContextMenuRegistry.RegistryItem = {
+      displayText: (scope) => {
+        return 'Copy (' + this.getPlatformPrefix() + ' + C)';
+      },
```

Comment:
* No need for curlies if body is a single return.
* Prefer template literals.
* On macOS the notation is "⌘C", not "⌘ + C", so any such notation should be included in the return value from `getPlatformPrefix`.

```suggestion
      displayText: (scope) => `Copy (${this.getPlatformPrefix()}C)`,
```

(Here and various places below.)

Reply:
Changed to the one-liner in all places.

Changed `getPlatformPrefix` to `return navigator.platform.startsWith('Mac') ? '⌘' : 'Ctrl + ';`

#155 includes the work for getting the prefix right for each platform.

@microbit-matt-hillsdon @kmcnaught Please confirm that those are the right prefixes for testing.